subsys/usb/class/netusb/function_rndis.c

```diff
@@ -0,0 +1,1018 @@
+		return rndis_keepalive_handle(data, len);
+	default:
+		SYS_LOG_ERR("Message 0x%x unhandled", msg->type);
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int rndis_send_media_status(u32_t media_status)
+{
+	struct rndis_media_status_indicate *ind;
+	struct net_buf *buf;
+
+	SYS_LOG_DBG("status %u", media_status);
+
+	buf = net_buf_alloc(&rndis_tx_pool, K_NO_WAIT);
+	if (!buf) {
+		SYS_LOG_ERR("Cannot get free buffer");
+		return -ENOMEM;
+	}
+
+	ind = net_buf_add(buf, sizeof(*ind));
+	ind->type = sys_cpu_to_le32(RNDIS_CMD_INDICATE);
+	ind->len = sys_cpu_to_le32(sizeof(*ind));
+
+	if (media_status) {
+		ind->status = sys_cpu_to_le32(RNDIS_STATUS_CONNECT_MEDIA);
+	} else {
+		ind->status = sys_cpu_to_le32(RNDIS_STATUS_DISCONNECT_MEDIA);
+	}
+
+	ind->buf_len = 0;
+	ind->buf_offset = 0;
+
+	rndis_queue_rsp(buf);
+
+	/* Nofity about ready reply */
+	rndis_notify_rsp();
+
+	return 0;
+}
+
+static int handle_encapsulated_rsp(u8_t **data, u32_t *len)
+{
+	struct net_buf *buf;
+
+	SYS_LOG_DBG("");
+
+	buf = net_buf_get(&rndis_tx_queue, K_NO_WAIT);
+	if (!buf) {
+		SYS_LOG_ERR("Error getting response buffer");
+		*len = 0;
+		return -ENODATA;
+	}
+
+	net_hexdump("RSP <", buf->data, buf->len);
+
+	memcpy(*data, buf->data, buf->len);
+	*len = buf->len;
+
+	net_buf_unref(buf);
+
+	return 0;
+}
+
+static int rndis_class_handler(struct usb_setup_packet *setup, s32_t *len,
+			       u8_t **data)
+{
+	SYS_LOG_DBG("");
+
+	if (setup->bRequest == CDC_SEND_ENC_CMD &&
+	    REQTYPE_GET_DIR(setup->bmRequestType) == REQTYPE_DIR_TO_DEVICE) {
+		/*
+		 * Instead of handling here, queue
+		 * handle_encapsulated_cmd(*data, *len);
+		 */
+		queue_encapsulated_cmd(*data, *len);
+	} else if (setup->bRequest == CDC_GET_ENC_RSP &&
+		   REQTYPE_GET_DIR(setup->bmRequestType) ==
+		   REQTYPE_DIR_TO_HOST) {
+		handle_encapsulated_rsp(data, len);
+	} else {
+		*len = 0; /* FIXME! */
+		SYS_LOG_WRN("Unknown USB packet req 0x%x type 0x%x",
+			    setup->bRequest, setup->bmRequestType);
+	}
+
+	return 0;
+}
+
+static void cmd_thread(void)
+{
+	SYS_LOG_INF("Command thread started");
+
+	while (true) {
+		struct net_buf *buf;
+
+		buf = net_buf_get(&rndis_cmd_queue, K_FOREVER);
+
+		SYS_LOG_DBG("got buf %p", buf);
+
+		handle_encapsulated_cmd(buf->data, buf->len);
+
+		net_buf_unref(buf);
+
+		k_yield();
+	}
+}
+
+/*
+ * RNDIS Send functions
+ */
+
+static u8_t *rndis_hdr_add(u8_t *buf, u32_t len)
+{
+	struct rndis_payload_packet *hdr = (void *)buf;
+	u8_t offset = (u8_t *)&hdr->payload_offset - (u8_t *)hdr;
+
+	memset(hdr, 0, sizeof(*hdr));
+
+	hdr->type = sys_cpu_to_le32(RNDIS_DATA_PACKET);
+	hdr->len = sys_cpu_to_le32(len + sizeof(*hdr));
+	hdr->payload_offset = sys_cpu_to_le32(sizeof(*hdr) - offset);
+	hdr->payload_len = sys_cpu_to_le32(len);
+
+	SYS_LOG_DBG("type %u len %u payload offset %u payload len %u",
+		    hdr->type, hdr->len, hdr->payload_offset, hdr->payload_len);
+
+	return buf + sizeof(*hdr);
+}
+
+static u8_t buf[RNDIS_GEN_MAX_TOTAL_SIZE];  /* FIXME */
```

Comment:
`buf` is not a good name for a global variable; it's in fact very easy to shadow it locally.  Please be more specific: `send_buffer` is preferable.

Also, is there a possibility that two threads will call `rndis_send()`?

In addition: what's that FIXME comment is referring to?

Reply:
About "(u8_t *)&hdr->payload_offset - (u8_t *)hdr)" construction: those are pointers to struct hdr, they should not be validated!
Now for some reason I cannot reply to all github comments separately. So how net_pkt_append_all() can be called if in_pkt == NULL?
For other comments I am rewriting send() function so those issues would be fixed.

Reply:
I also have a check: 
if (buf_len < sizeof(*hdr)) {
		SYS_LOG_ERR("Too small packet len %u", buf_len);
		return -EINVAL;
	}